op-service/eth/ether.go

```diff
@@ -27,3 +30,260 @@ func GweiToWei(gwei float64) (*big.Int, error) {
+// This type is not optimized for speed,
+// but is instead designed for readability and to remove mutability foot-guns.
+type ETH uint256.Int
+
+// String prints the amount of ETH, with thousands comma-separators, and unit.
+// If the amount is perfectly divisible by 1 ether, the amount is printed in ethers.
+// If the amount is perfectly divisible by 1 gwei, the amount is printed in gwei.
+// If not neatly divisible, the amount is printed in wei.
+// This String function is optimized for readability, without precision loss,
+// to not hide any precision data during debugging.
+func (e ETH) String() string {
+	vWei := (*uint256.Int)(&e)
+	if vWei.Sign() == 0 {
+		return "0 wei"
+	}
+	var vGWei uint256.Int
+	var remainder uint256.Int
+	vGWei.DivMod(vWei, weiPerGWei, &remainder)
+	if remainder.Sign() == 0 {
+		// an exact number of gwei
+		var vEth uint256.Int
+		vEth.DivMod(vWei, weiPerEth, &remainder)
+		if remainder.Sign() == 0 {
+			// an exact number of eth
+			return vEth.PrettyDec(',') + " ether"
+		}
+		return vGWei.PrettyDec(',') + " gwei"
+	}
+	return vWei.PrettyDec(',') + " wei"
+}
+
+// Decimal returns the amount, in wei, in decimal form.
+func (e ETH) Decimal() string {
+	return (*uint256.Int)(&e).String()
+}
+
+// Hex returns the amount, in wei, in hexadecimal form with 0x prefix.
+func (e ETH) Hex() string {
+	return (*uint256.Int)(&e).Hex()
+}
+
+// Format implements fmt.Formatter
+func (e ETH) Format(s fmt.State, ch rune) {
+	(*uint256.Int)(&e).Format(s, ch)
+}
+
+// WeiFloat returns the amount as floating point number, in wei (approximate).
+// Warning: precision loss. This may not present the exact number of wei.
+func (e ETH) WeiFloat() float64 {
+	vWei := (*uint256.Int)(&e)
+	return vWei.Float64()
+}
+
+// EtherString returns the amount, string-ified, forced in ether units (excl. unit suffix)
+func (e ETH) EtherString() string {
+	var ethers uint256.Int
+	var remainder uint256.Int
+	ethers.DivMod((*uint256.Int)(&e), OneEther.ToU256(), &remainder)
+	if remainder.Sign() == 0 {
+		return ethers.Dec()
+	}
+	// No trailing zeroes
+	suffix := strings.TrimRight(fmt.Sprintf("%018d", &remainder), "0")
+	return ethers.Dec() + "." + suffix
+}
+
+// ToBig converts to *big.Int, in wei.
+func (e ETH) ToBig() *big.Int {
+	return (*uint256.Int)(&e).ToBig()
+}
+
+// SetFromUint64 sets from an uint64, in wei.
+// The upper 192 bits are zeroed.
+func (e *ETH) SetFromUint64(v uint64) {
+	(*uint256.Int)(e).SetUint64(v)
+}
+
+// SetFromBig sets from a *big.Int,
+// and returns true if the value does not fit and overflows the uint256 max size.
+func (e *ETH) SetFromBig(v *big.Int) (overflow bool) {
+	if v == nil {
+		panic("cannot set to nil big.Int")
+	}
+	return (*uint256.Int)(e).SetFromBig(v)
+}
+
+// SetFromUint256 sets from a *uint256.Int
+func (e *ETH) SetFromUint256(v *uint256.Int) {
+	if v == nil {
+		panic("cannot set to nil uint256.Int")
+	}
+	*e = ETH(*v)
+}
+
+// Set sets to the given ETH value
+func (e *ETH) Set(v ETH) {
+	*e = v
+}
```

Comment:
I'm somewhat confused about these methods given we were aiming for immutability. Is there a particular reason to have these methods or could we go fully immutable and just have `func FromBig(v *big.Int) ETH` kind of methods.